internal/util/password/scramsha256.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package password
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/xdg-go/scram"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// NewSCRAMSHA256 authentication.
+func NewSCRAMSHA256(username, password string) (*types.Document, error) {
+	client, err := scram.SHA256.NewClient(username, password, "")
+	if err != nil {
+		return nil, err
+	}
+
+	salt := make([]byte, 45)
+	if _, err := rand.Read(salt); err != nil {
+		return nil, fmt.Errorf("failed to generate salt: %w", err)
+	}
+
+	kf := scram.KeyFactors{
+		Salt:  base64.StdEncoding.EncodeToString(salt),
```

Comment:
That should not be [encoded](https://pkg.go.dev/github.com/xdg-go/scram@v1.1.2#KeyFactors).

```suggestion
		Salt:  string(salt),
```

That will fix the invalid client [proof](https://github.com/xdg-go/scram/blob/v1.1.2/server_conv.go#L143).

Reply:
Without actually looking into it, may I suggest checking if `string` BSON type is not confused with `types.Binary`?